lib/eig_block_trlm.cpp

```diff
@@ -0,0 +1,771 @@
+      }
+    }
+
+    profile.TPSTOP(QUDA_PROFILE_EIGEN);
+  }
+
+  void BLKTRLM::computeBlockKeptRitz(std::vector<ColorSpinorField *> &kSpace)
+  {
+    int offset = nKr + block_size;
+    int dim = nKr - num_locked;
+
+    // Multi-BLAS friendly array to store part of Ritz matrix we want
+    Complex *ritz_mat_keep = (Complex *)safe_malloc((dim * iter_keep) * sizeof(Complex));
+
+    // If we have memory availible, do the entire rotation
+    if (batched_rotate <= 0 || batched_rotate >= iter_keep) {
+      if ((int)kSpace.size() < offset + iter_keep) {
+        ColorSpinorParam csParamClone(*kSpace[0]);
+        csParamClone.create = QUDA_ZERO_FIELD_CREATE;
+        if (getVerbosity() >= QUDA_VERBOSE) printfQuda("Resizing kSpace to %d vectors\n", offset + iter_keep);
+        kSpace.reserve(offset + iter_keep);
+        for (int i = kSpace.size(); i < offset + iter_keep; i++) {
+          kSpace.push_back(ColorSpinorField::Create(csParamClone));
+        }
+      }
+
+      // Pointers to the relevant vectors
+      std::vector<ColorSpinorField *> vecs_ptr;
+      std::vector<ColorSpinorField *> kSpace_ptr;
+
+      // Alias the extra space vectors, zero the workspace
+      kSpace_ptr.reserve(iter_keep);
+      for (int i = 0; i < iter_keep; i++) {
+        kSpace_ptr.push_back(kSpace[offset + i]);
+        blas::zero(*kSpace_ptr[i]);
+      }
+
+      // Alias the vectors we wish to keep, populate the Ritz matrix and transpose.
+      vecs_ptr.reserve(dim);
+      for (int j = 0; j < dim; j++) {
+        vecs_ptr.push_back(kSpace[num_locked + j]);
+        for (int i = 0; i < iter_keep; i++) { ritz_mat_keep[j * iter_keep + i] = block_ritz_mat[i * dim + j]; }
+      }
+
+      // multiBLAS caxpy
+      blas::caxpy(ritz_mat_keep, vecs_ptr, kSpace_ptr);
+
+      // Copy compressed Krylov
+      for (int i = 0; i < iter_keep; i++) std::swap(kSpace[i + num_locked], kSpace[offset + i]);
+
+    } else {
+
+      // Do batched rotation to save on memory
+      int batch_size = batched_rotate;
+      int full_batches = iter_keep / batch_size;
+      int batch_size_r = iter_keep % batch_size;
+      bool do_batch_remainder = (batch_size_r != 0 ? true : false);
+
+      if ((int)kSpace.size() < offset + batch_size) {
+        ColorSpinorParam csParamClone(*kSpace[0]);
+        csParamClone.create = QUDA_ZERO_FIELD_CREATE;
+        if (getVerbosity() >= QUDA_VERBOSE) printfQuda("Resizing kSpace to %d vectors\n", offset + batch_size);
+        kSpace.reserve(offset + batch_size);
+        for (int i = kSpace.size(); i < offset + batch_size; i++) {
+          kSpace.push_back(ColorSpinorField::Create(csParamClone));
+        }
+      }
+
+      profile.TPSTART(QUDA_PROFILE_EIGEN);
+      MatrixXcd mat = MatrixXcd::Zero(dim, iter_keep);
+      for (int j = 0; j < iter_keep; j++)
+        for (int i = 0; i < dim; i++) mat(i, j) = block_ritz_mat[j * dim + i];
+
+      FullPivLU<MatrixXcd> matLU(mat);
```

Comment:
I can't see anything in the Eigen documentation about doing LU via SVD. Full pivoting is done because we need not only the P permutation matrix, but also the Q permutation matrix. Partial pivoting provides P only:

https://eigen.tuxfamily.org/dox/group__LU__Module.html

Reply:
Gotcha, my misunderstanding.